README.md
# X25519KeyAgreementKey2019 _(@digitalbazaar/x25519-key-agreement-key-2019)_

> An X25519 (Curve25519) DH (Diffie-Hellman) key implementation to work with the crypto-ld LDKeyPair API.

## Table of Contents

- [Security](#security)
- [Background](#background)
- [Install](#install)
- [Usage](#usage)
- [Contribute](#contribute)
- [Commercial Support](#commercial-support)
- [License](#license)

## Security

TBD

## Background

For use with [`crypto-ld`](https://github.com/digitalbazaar/crypto-ld) `>= 5.0`.

To actually perform encryption with those keys, we recommend you use
the [`minimal-cipher`](https://github.com/digitalbazaar/minimal-cipher) library.

This is a low-level level library to generate and serialize X25519 (Curve25519)
key pairs (uses `nacl.box` under the hood).

See also (related specs):

* [Linked Data Proofs](https://w3c-ccg.github.io/ld-proofs/)
* [Linked Data Cryptographic Suite Registry](https://w3c-ccg.github.io/ld-cryptosuite-registry/)

## Install

Requires Node.js 12+

To install locally (for development):

```
git clone https://github.com/digitalbazaar/x25519-key-agreement-key-2019.git
cd x25519-key-agreement-key-2019
npm install
```

## Usage

Importing:

```
const {X25519KeyAgreementKey2019} = require('@digitalbazaar/x25519-key-agreement-key-2019');

// Or, if you're testing code in the interactive Node CLI, right in this repo:
const {X25519KeyAgreementKey2019} = require('./');
```

Generating:

```js
const keyPair = await X25519KeyAgreementKey2019.generate({
  controller: 'did:example:1234'
});
```

Serializing just the public key:

```js
keyPair.export({publicKey: true});
// ->
{
  id: 'did:example:1234#z6LSbh9HiAU2zzBdFMdKZGHfg1UjvAYF8C8kYnkfGKuCxYEB',
  type: 'X25519KeyAgreementKey2019',
  controller: 'did:example:1234',
  publicKeyBase58: '1y8BrfAuXTt9yFZ2cmiMRGG5218Raxbfp2ymsFgFATR'
}
```

Serializing both the private and public key:

```js
// a different key pair than the previous example
await keyPair.export({publicKey: true, privateKey: true})
// ->
 {
  id: 'did:example:1234#z6LSjeJZaUHMvEKW7tEJXV4PrSm61NzxxHhDXF6zHnVtDu9g',
  type: 'X25519KeyAgreementKey2019',
  controller: 'did:example:1234',
  publicKeyBase58: '8y8Q4AUVpmbm2VrXzqYSXrYcAETrFgX4eGPJoKrMWXNv',
  privateKeyBase58: '95tmYuhqSuJqY77FEg78Zy3LFQ1cENxGv2wMvayk7Lqf'
}
```

Deserializing:

```js
// Loading public key only
const keyPair = await X25519KeyAgreementKey2019.from({
  id: 'did:example:1234#z6LSjeJZaUHMvEKW7tEJXV4PrSm61NzxxHhDXF6zHnVtDu9g',
  type: 'X25519KeyAgreementKey2019',
  controller: 'did:example:1234',
  publicKeyBase58: '8y8Q4AUVpmbm2VrXzqYSXrYcAETrFgX4eGPJoKrMWXNv'
});
```

## Contribute

See [the contribute file](https://github.com/digitalbazaar/bedrock/blob/master/CONTRIBUTING.md)!

PRs accepted.

If editing the Readme, please conform to the
[standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## Commercial Support

Commercial support for this library is available upon request from
Digital Bazaar: support@digitalbazaar.com

## License

[New BSD License (3-clause)](LICENSE) © Digital Bazaar


CHANGELOG.md
# @digitalbazaar/x25519-key-agreement-key-2019 ChangeLog

## 6.0.0 - 2022-06-02

### Changed
- **BREAKING**: Convert to module (ESM).
- **BREAKING**: Require Node.js >=14.
- Update dependencies.
- Lint module.

## 5.2.0 - 2022-05-06

### Changed
- Use `@noble/ed25519` to convert public ed25519 keys to x25519.

### Fixed
- Fix broken ed25519 2020 conversion code.

## 5.1.1 - 2021-05-25

### Changed
- **BREAKING**: Point suite context to its own standalone url (not security/v2).

## 5.1.0 - 2021-04-02

### Added
- Add `revoked` export tests. (To support `CryptoLD`'s new `fromKeyId()`
  method.) Also add `includeContext` flag to `export()`.

## 5.0.1 - 2021-03-25

## Changed
- Remove `env.js`, switch to our usual node/browser setup. Should fix webpack
  problems downstream.

## 5.0.0 - 2021-03-17

## Changed
- Update to use `crypto-ld v5.0`.
- **BREAKING**: Removed helper methods `addPublicKey` and `addPrivateKey`.

## 4.1.0 - 2021-03-14

### Added
- `fromEdKeyPair()` is now an alias for `fromEd25519VerificationKey2018()` to
  maintain backwards compatibility. New code should use
  `fromEd25519VerificationKey2020()` (or whatever the latest Ed25519 suite is).

## 4.0.0 - 2021-03-11

### Changed
- **BREAKING**: Rename repo and NPM package name to
  `@digitalbazaar/x25519-key-agreement-key-2019`.
- **BREAKING**: Rename `addEncodedPublicKey()` to `addPublicKey()`.
- **BREAKING**: Rename `addEncryptedPrivateKey()` to `addPrivateKey()`.
- **BREAKING**: Changed `verifyFingerprint()` param signature to use named
  params.
- **BREAKING**: Changed `fromEdKeyPair()` param signature to use named params.
- **BREAKING**: Changed `convertFromEdPublicKey()` param signature to use named
  params.
- **BREAKING**: Changed `convertFromEdPrivateKey()` param signature to use named
  params.
- See also [`crypto-ld` v4.0 Changelog](https://github.com/digitalbazaar/crypto-ld/blob/master/CHANGELOG.md#400---2020-08-01)

### Purpose and Upgrade Instructions
See [`crypto-ld` v4.0 Purpose](https://github.com/digitalbazaar/crypto-ld/blob/master/CHANGELOG.md#400---purpose)
and [`crypto-ld` Upgrade from v3.7 notes](https://github.com/digitalbazaar/crypto-ld/blob/master/CHANGELOG.md#upgrading-from-v370)

## 3.1.0 - 2020-10-08

### Changes
- Use node-forge@0.10.0.
- Update dev dependencies.

## 3.0.0 - 2020-08-01

### Added
- Auto-initialize key.id based on controller (if it's present).

### Changed
- **BREAKING**: Explicitly make `publicBase58` property required for Ed25519
  type keys (throw error if missing).

## 2.0.0 - 2020-03-09

### Changed
- **BREAKING**: Changed the key fingerprint prefix to the recently registered
  `multicodec` value of `0xec`.

## 1.0.0 - 2020-02-25

- See git history for changes.


lib/X25519KeyPair.js
/*!
 * Copyright (c) 2019 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

import {util} from 'node-forge';
const base58 = util.binary.base58;
import {LDKeyPair} from 'crypto-ld';
import {base58Decode} from 'crypto-ld/lib/util';
import ed2curve from 'ed2curve';
import nacl from 'tweetnacl';

const ed25519_to_x25519 = require('ed25519-to-x25519.wasm');
const {promisify} = require('util');
const ready = promisify(ed25519_to_x25519.ready);

export class X25519KeyPair extends LDKeyPair {
  /**
   * An implementation of x25519
   * [X25519 Key Agreement 2019]{@link https://w3c-dvcg.github.io/}
   * representation
   * @example
   * > const privateKeyBase58 =
   *   '...';
   * > const options = {
   *   publicKeyBase58: '...',
   *   privateKeyBase58
   * };
   * > const DHKey = new X25519KeyPair(options);
   * > DHKey
   * X25519KeyPair { ...
   * @param {KeyPairOptions} options - Base58 keys plus other options
   * @param {string} options.controller
   * @param {string} options.id
   * @param {string} options.publicKeyBase58 - Base58 encoded Public Key
   * unencoded is 32-bytes.
   * @param {string} options.privateKeyBase58 - Base58 Private Key
   * unencoded is 32-bytes.
   */
  constructor(options = {}) {
    super(options);
    this.type = 'X25519KeyAgreementKey2019';
    this.privateKeyBase58 = options.privateKeyBase58;
    this.publicKeyBase58 = options.publicKeyBase58;
  }
  /**
   * Returns the Base58 encoded public key.
   * @implements {LDKeyPair#publicKey}
   * @readonly
   *
   * @returns {string} The Base58 encoded public key.
   */
  get publicKey() {
    return this.publicKeyBase58;
  }
  /**
   * Returns the Base58 encoded private key.
   * @implements {LDKeyPair#privateKey}
   * @readonly
   *
   * @returns {string} The Base58 encoded private key.
   */
  get privateKey() {
    return this.privateKeyBase58;
  }

  /**
   * Generates a new public/private X25519 Key Pair.
   * Note: This is async only to match the async signature of other LDKeyPair
   * subclasses.
   * @example
   * > const keyPair = await X25519KeyPair.generate();
   * > keyPair
   * X25519KeyPair { ...
   * @param {KeyPairOptions} [options={}]
   * @param {string} [options.controller]
   * @param {string} [options.id]
   *
   * @returns {Promise<X25519KeyPair>} Generates a key pair.
   */
  static async generate(options = {}) {
    // Each is a Uint8Array with 32-byte key
    const {publicKey, secretKey: privateKey} = nacl.box.keyPair();

    return new X25519KeyPair({
      publicKeyBase58: base58.encode(publicKey),
      privateKeyBase58: base58.encode(privateKey),
      ...options
    });
  }

  /**
   * Creates an X25519KeyPair Key Pair from an existing key (constructor
   * method).
   * @example
   * > const options = {
   *   id,
   *   controller,
   *   publicKeyBase58,
   *   privateKeyBase58: privateKey
   * };
   * > const key = await X25519KeyPair.from(options);
   * > key
   * X25519KeyPair { ...
   * @param {Object} options
   * @param {string} [options.privateKeyBase58] - A Base58 encoded Private key
   *
   * @returns {X25519KeyPair} An X25519 Key Pair.
   */
  static async from(options) {
    return new X25519KeyPair(options);
  }

  static async fromEdKeyPair(edKeyPair) {
    const xKey = new X25519KeyPair({
      controller: edKeyPair.controller
    });

    xKey.publicKeyBase58 = await X25519KeyPair
      .convertFromEdPublicKey(edKeyPair.publicKeyBase58);

    if(edKeyPair.privateKeyBase58) {
      xKey.privateKeyBase58 = X25519KeyPair
        .convertFromEdPrivateKey(edKeyPair.privateKeyBase58);
    }

    return xKey;
  }

  /**
   * @param {string} edPublicKeyBase58 - base58 encoded Ed25519 Public key
   *
   * @returns {string} base58 encoded X25519 Public key.
   */
  static async convertFromEdPublicKey(edPublicKeyBase58) {
    const edPubkeyBytes = base58Decode({
      decode: base58.decode,
      keyMaterial: edPublicKeyBase58,
      type: 'public'
    });

    // Converts a 32-byte Ed25519 public key into a 32-byte Curve25519 key
    // Returns null if the given public key in not a valid Ed25519 public key.

    // const dhPubkeyBytes = ed2curve.convertPublicKey(edPubkeyBytes);

    // FIXME: there should be check to ensure WASM is ready?
    // ed25519_to_x25519 has a `ready` helper with a callback for this purpose
    // is this still required in modern nodejs?
    try {
      await ready();
    } catch(e) {
      // ready callback is not compatible with promisify so ready is indicated
      // by it throwing here
      console.log('wasm ready');
    }

    const dhPubkeyBytes = ed25519_to_x25519.convert_public_key(edPubkeyBytes);

    if(!dhPubkeyBytes) {
      throw new Error(
        'Error converting to X25519; Invalid Ed25519 public key.');
    }
    const dhPublicKeyBase58 = base58.encode(dhPubkeyBytes);
    return dhPublicKeyBase58;
  }

  /**
   * @param {string} edPrivateKeyBase58 - base58 encoded Ed25519 Private key
   *
   * @returns {string} base58 encoded X25519 Private key.
   */
  static convertFromEdPrivateKey(edPrivateKeyBase58) {
    const edPrivkeyBytes = base58Decode({
      decode: base58.decode,
      keyMaterial: edPrivateKeyBase58,
      type: 'private'
    });
    // Converts a 64-byte Ed25519 secret key (or just the first 32-byte part of
    // it, which is the secret value) into a 32-byte Curve25519 secret key
    const dhPrivkeyBytes = ed2curve.convertSecretKey(edPrivkeyBytes);
    if(!dhPrivkeyBytes) {
      throw new Error(
        'Error converting to X25519; Invalid Ed25519 private key.');
    }
    const dhPrivateKeyBase58 = base58.encode(dhPrivkeyBytes);
    return dhPrivateKeyBase58;
  }

  /**
   * Adds a public key base to a public key node.
   * Used by LDKeyPair.publicNode(), to serialize the public key material.
   * @example
   * > keyPair.publicNode();
   * {
   *   id: 'did:example:1234#z6LSbh9HiAU2zzBdFMdKZGHfg1UjvAYF8C8kYnkfGKuCxYEB',
   *   type: 'X25519KeyAgreementKey2019',
   *   controller: 'did:example:1234',
   *   publicKeyBase58: '1y8BrfAuXTt9yFZ2cmiMRGG5218Raxbfp2ymsFgFATR'
   * }
   * @param {object} publicKeyNode - The public key node
   * @param {string} publicKeyNode.publicKeyBase58 - Base58 Public Key
   *
   * @returns {object} A PublicKeyNode in a block.
   */
  addEncodedPublicKey(publicKeyNode) {
    publicKeyNode.publicKeyBase58 = this.publicKeyBase58;
    return publicKeyNode;
  }

  /**
   * Adds an encrypted private key to the KeyPair.
   * Used by LDKeyPair.export(), to serialize public + private key pair.
   *
   * Usage:
   *
   * ```
   * await keyPair.export();
   * // ->
   * {
   *   id: 'did:example:1234#z6LSjeJZaUHMvEKW7tEJXV4PrSm61NzxxHhDXF6zHnVtDu9g',
   *   type: 'X25519KeyAgreementKey2019',
   *   controller: 'did:example:1234',
   *   publicKeyBase58: '8y8Q4AUVpmbm2VrXzqYSXrYcAETrFgX4eGPJoKrMWXNv',
   *   privateKeyBase58: '95tmYuhqSuJqY77FEg78Zy3LFQ1cENxGv2wMvayk7Lqf'
   * }
   * ```
   *
   * @param {object} keyNode - A plain object.
   *
   * @return {object} The keyNode with an encrypted private key attached.
   */
  async addEncryptedPrivateKey(keyNode) {
    if(this.passphrase !== null) {
      throw new Error('Encrypted export not yet implemented.');
    }
    // no passphrase, do not encrypt private key
    keyNode.privateKeyBase58 = this.privateKeyBase58;
    return keyNode;
  }

  /**
   * Generates and returns a multiformats encoded
   * X25519 public key fingerprint (for use with cryptonyms, for example).
   * @see https://github.com/multiformats/multicodec
   *
   * @param {string} publicKeyBase58 - The base58 encoded public key material.
   *
   * @returns {string} The fingerprint.
   */
  static fingerprintFromPublicKey({publicKeyBase58}) {
    // X25519 cryptonyms are multicodec encoded values, specifically:
    // (multicodec('x25519-pub') + key bytes)
    const pubkeyBytes = base58Decode({
      decode: base58.decode,
      keyMaterial: publicKeyBase58,
      type: 'public'
    });
    const buffer = new Uint8Array(2 + pubkeyBytes.length);
    // See https://github.com/multiformats/multicodec/blob/master/table.csv
    // 0xec is the value for X25519 public key
    // 0x01 is from varint.encode(0xec) -> [0xec, 0x01]
    // See https://github.com/multiformats/unsigned-varint
    buffer[0] = 0xec; //
    buffer[1] = 0x01;
    buffer.set(pubkeyBytes, 2);
    // prefix with `z` to indicate multi-base base58btc encoding
    return `z${base58.encode(buffer)}`;
  }

  /**
   * Creates an instance of X25519KeyPair from a key fingerprint.
   *
   * @param {string} fingerprint
   * @returns {X25519KeyPair}
   * @throws Unsupported Fingerprint Type.
   */
  static fromFingerprint({fingerprint}) {
    // skip leading `z` that indicates base58 encoding
    const buffer = base58.decode(fingerprint.substr(1));

    // buffer is: 0xec 0x01 <public key bytes>
    if(buffer[0] !== 0xec || buffer[1] !== 0x01) {
      throw new Error(`Unsupported Fingerprint Type: ${fingerprint}`);
    }

    return new X25519KeyPair({
      publicKeyBase58: base58.encode(buffer.slice(2))
    });
  }

  /**
   * Derives a shared secret via a given public key, typically for use
   * as one parameter for computing a shared key. It should not be used as
   * a shared key itself, but rather input into a key derivation function (KDF)
   * to produce a shared key.
   * @param {LDKeyPair} publicKey - Remote key pair
   * @throws {TypeError} On invalid base58 encoding of public or private keys
   * @returns {String}
   */
  deriveSecret({publicKey}) {
    const remotePubkeyBytes = base58Decode({
      decode: base58.decode,
      keyMaterial: publicKey.publicKeyBase58,
      type: 'public'
    });

    const privateKeyBytes = base58Decode({
      decode: base58.decode,
      keyMaterial: this.privateKeyBase58,
      type: 'private'
    });

    return nacl.scalarMult(privateKeyBytes, remotePubkeyBytes);
  }

  /**
   * Generates and returns a multiformats encoded
   * X25519 public key fingerprint (for use with cryptonyms, for example).
   * @see https://github.com/multiformats/multicodec
   *
   * @returns {string} The fingerprint.
   */
  fingerprint() {
    const {publicKeyBase58} = this;
    return X25519KeyPair.fingerprintFromPublicKey({publicKeyBase58});
  }

  /**
   * Tests whether the fingerprint was generated from a given key pair.
   * @example
   * > xKeyPair.verifyFingerprint('...');
   * {valid: true};
   * @param {string} fingerprint - A Base58 public key.
   *
   * @returns {Object} An object indicating valid is true or false.
   */
  verifyFingerprint(fingerprint) {
    // fingerprint should have `z` prefix indicating
    // that it's multi-base encoded
    if(!(typeof fingerprint === 'string' && fingerprint[0] === 'z')) {
      return {
        error: new Error('`fingerprint` must be a multibase encoded string.'),
        valid: false
      };
    }
    let fingerprintBuffer;
    try {
      fingerprintBuffer = base58Decode({
        decode: base58.decode,
        keyMaterial: fingerprint.slice(1),
        type: `fingerprint's`
      });
    } catch(e) {
      return {error: e, valid: false};
    }
    let publicKeyBuffer;
    try {
      publicKeyBuffer = base58Decode({
        decode: base58.decode,
        keyMaterial: this.publicKeyBase58,
        type: 'public'
      });
    } catch(e) {
      return {error: e, valid: false};
    }
    // validate the first buffer multicodec bytes 0xec 0x01
    const valid = fingerprintBuffer[0] === 0xec &&
      fingerprintBuffer[1] === 0x01 &&
      publicKeyBuffer.toString() === fingerprintBuffer.slice(2).toString();
    if(!valid) {
      return {
        error: new Error('The fingerprint does not match the public key.'),
        valid: false
      };
    }
    return {valid};
  }
}
